### PR TITLE
Filter Ads accounts

### LINKED ISSUE #317: Filter ads accounts
Currently, we list all existing ads account to a user when connecting, including Manager and test accounts.

I'd need to double-check with @JamesAntisdel but I think I recall briefly discussing that we should filter the list.

Screen Shot 2021-03-10 at 6 57 05 pm

(if possible)

### PR DESCRIPTION
**Pull request by mikkamp**
### Changes proposed in this Pull Request:

This PR filters Ads accounts to exclude any test / manager accounts. This is done by sending a separate request to retrieve the account details using a search request. The calls can't be batched because each one needs to send a different account ID. Since we retrieve account details we now also have access to the account name which is included in the response. This changes the responses from an array of integers to an array of account details.

Note: We use search requests instead of GET requests because the latter are limited to a 1000 per day.
Request the account currency was also changed in this PR (although that gets cached so isn't repeated).

When testing I did notice that it's a bit slower and will increase the amount of requests that are sent (one for all account IDs and a request per account). However this is only needed when onboarding/connecting an Ads account, so I think it's still acceptable.

The issue mentions checking if we have ADMIN access to the account which would allow us to link it. In order to do so we would need to expand the query to retrieve data from both `customer` and `customer_user_access`. I haven't included this for now as it could potentially slow down the requests even more.

Closes #317

#### Get accounts
`GET https://domain.test/wp-json/wc/gla/ads/accounts`

Response:
```json
[
	{
		"id": 1234567890,
		"name": "Main Ads Account"
	},
	{
		"id": 2345678901,
		"name": "Secondary Ads Account"
	}
]
```

If you have access to any test or manager accounts, they should no longer show up in the list.

### TODO:
- [x] UI still expects a list of account ID's, so it will need to be adjusted to support the new account details.

### Changelog entry
* Add - Filter Ads accounts to exclude manager and test accounts.
* Add - Return account names when retrieving the list of existing accounts.

### CODE REVIEW

**Comment by jorgemd24** [comment]
This is a follow-up of pcTzPl-EY-p2#comment-1319

I did a quick testing and I saw that the manager and test accounts are filtered (without checking the access role). If we go ahead with this PR I can do the review and adjust the UI so it is able to work with the new response, a JS dev can review this part.

As far as I know, there is only one place where it displays the existing account ads. See image:

image

In this PR, the accounts are displayed with the following format `accountID : accountName` in the connection test page. Should we follow the same format for the `Select an existing account` section?

https://github.com/woocommerce/google-listings-and-ads/blob/02f4aa3e30d0458f96eb9f696d66c728df02052a/src/ConnectionTest.php#L914

image

cc: @jconroy, @mikkamp, @j111q

**Comment by mikkamp** [comment]
> In this PR, the accounts are displayed with the following format accountID : accountName in the connection test page. Should we follow the same format for the Select an existing account section?

For selecting the Merchant Center account we use the [following format](https://github.com/woocommerce/google-listings-and-ads/blob/1.13.6/js/src/components/merchant-center-select-control/index.js#L19-L20):

```
Account name ・ https://domain.com (123456789)
```

For the Ads account we only have the name and ID, but it would seem more consistent to go with something like:
```
Account name (123456789)
```

File: src/API/Google/Ads.php
Comment (jorgemd24) [inline src/API/Google/Ads.php]: 💅

What do you think if we rename the variable $return for $accounts?

File: src/API/Google/Ads.php
Comment (jorgemd24) [inline src/API/Google/Ads.php]: It seems that `CustomerServiceClient` is not used, we could remove it.

File: src/API/Google/Ads.php
Comment (jorgemd24) [inline src/API/Google/Ads.php:157]: Not introduced in this PR but maybe is a good moment to update it. The description in this function seems incorrect `Check if we have access to the merchant account.`, I think it should be `Check if we have access to the ads account.`

File: tests/Unit/API/Site/Controllers/Ads/AccountControllerTest.php
Comment (jorgemd24) [inline tests/Unit/API/Site/Controllers/Ads/AccountControllerTest.php]: We could update no_ids with no_accounts, but maybe it sounds too repetive

**Review by jorgemd24 — COMMENTED** [review]
Thanks, @mikkamp, I tested it locally and it works as expected 👍 

Something that I noticed is that the schema response for the endpoint `ads/accounts` does not match with the actual response:

https://github.com/woocommerce/google-listings-and-ads/blob/29ea3fbad09ec7edb5d9464b6ef803b9a965b181/src/API/Site/Controllers/Ads/AccountController.php#L170-L186
Actual response: 

```
	[
	   {
		"id": 1234567890,
		"name": "Main Ads Account"
	      }
	]
```

File: js/src/components/google-ads-account-card/connect-ads/ads-account-select-control.js
Comment (mikkamp) [inline js/src/components/google-ads-account-card/connect-ads/ads-account-select-control.js:10]: As far as I can see the function toAccountText is no longer used after this removal, should we remove that from the utils directory as well?

**Comment by mikkamp** [comment]
> Something that I noticed is that the schema response for the endpoint ads/accounts does not match with the actual response

This is one of those cases where the controller has multiple schema responses. The one that's setup is the response for creating/linking an ads account. When we list accounts the response is not passed through the function `prepare_item_for_response`. Ideally we should discuss these scenarios a bit more in pb0Spc-2pd-p2.

**Comment by mikkamp** [comment]
Thanks for the review @jorgemd24
I've address those points you mentioned, it's ready for another round. In regards to the JS code you added, it's working fine and looks good to me. But feel free to get a second opinion if you think it's necessary.

**Review by jorgemd24 — APPROVED** [review]
Thanks, @mikkamp for the adjustments, it is looking good. I am not sure if you missed this comment https://github.com/woocommerce/google-listings-and-ads/pull/1390#discussion_r909514468 or do you prefer to name the variable `$return` rather than `$accounts`?

>This is one of those cases where the controller has multiple schema responses. The one that's setup is the response for creating/linking an ads account. When we list accounts the response is not passed through the function prepare_item_for_response. Ideally we should discuss these scenarios a bit more in pb0Spc-2pd-p

Yes, I think adds confusion for the consumer as the schema is not telling us how our data is structured. 

>In regards to the JS code you added, it's working fine and looks good to me. But feel free to get a second opinion if you think it's necessary.

Thanks, I think it is fine, basically, I am following the same code as [MerchantCenterSelectControl](https://github.com/woocommerce/google-listings-and-ads/blob/1.13.6/js/src/components/merchant-center-select-control/index.js#L19-L20)

**Comment by mikkamp** [comment]
Thanks for the additional review.

> I am not sure if you missed this comment https://github.com/woocommerce/google-listings-and-ads/pull/1390#discussion_r909514468 or do you prefer to name the variable $return rather than $accounts?

I read it but missed it when I went to apply all the fixes. It was copied from some older code and we use that pattern in a few other places in the code, but I agree that it's much clearer to specify what the variable contains.

Everything looks good to go then, I'll merge it once the tests pass. Thanks again.

**Pull request merged**
